### PR TITLE
replace badger version with transaction TS.

### PR DESCRIPTION
**Pull request by coocood**
- eliminate the extra write cost caused by move old version data to another key.
- Simplify the older version data management logic.

### CODE REVIEW

**Comment by coocood** [comment]
@bobotu PTAL

File: tikv/raftstore/engine.go [hunk 1]
```diff
@@ -165,53 +166,65 @@ func (wb *WriteBatch) Len() int {
...
 	wb.entries = append(wb.entries, &badger.Entry{
 		Key:      key,
 		Value:    val,
 		UserMeta: useMeta,
 	})
-	wb.size += len(key) + len(val) + len(useMeta)
+	wb.size += key.Len() + len(val) + len(useMeta)
+}
+
+func (wb *WriteBatch) SetOpLock(key y.Key, useMeta []byte) {
```
Comment (bobotu) [inline tikv/raftstore/engine.go:209]: s/useMeta/userMeta/

File: tikv/raftstore/engine.go [hunk 2]
```diff
@@ -165,53 +166,65 @@ func (wb *WriteBatch) Len() int {
...
+func (wb *WriteBatch) Rollback(key y.Key) {
+	rollbackKey := mvcc.EncodeExtraTxnStatusKey(key.UserKey, key.Version)
+	wb.entries = append(wb.entries, &badger.Entry{
+		Key:      y.KeyWithTs(rollbackKey, key.Version),
+		UserMeta: mvcc.NewDBUserMeta(key.Version, 0),
 	})
 }
 
-func (wb *WriteBatch) SetWithUserMeta(key, val, useMeta []byte) {
+func (wb *WriteBatch) SetWithUserMeta(key y.Key, val, useMeta []byte) {
```
Comment (bobotu) [inline tikv/raftstore/engine.go:200]: s/useMeta/userMeta/

File: tikv/mvcc.go [hunk 3]
```diff
@@ -274,6 +280,30 @@ func (store *MVCCStore) PessimisticLock(reqCtx *requestCtx, req *kvrpcpb.Pessimi
 	return nil, err
 }
 
+// extraTxnStatus can be rollback or Op_Lock that only contains transaction status info, no values.
+type extraTxnStatus struct {
+	commitTS   uint64
+	isRollback bool
```
[the remark this answers is not in the document]
Comment (coocood) [reply tikv/mvcc.go:283]: No, if the status is not found, isRollback is false and commitTS is 0.

**Pull request merged**